service/storage_service.cc

```diff
@@ -4412,7 +4412,50 @@ future<> storage_service::move(token new_token) {
 
 future<std::vector<storage_service::token_range_endpoints>>
 storage_service::describe_ring(const sstring& keyspace, bool include_only_local_dc) const {
-    return locator::describe_ring(_db.local(), _gossiper, keyspace, include_only_local_dc);
+    if (_db.local().find_keyspace(keyspace).get_replication_strategy().uses_tablets()) {
+        throw std::runtime_error(format("The keyspace {} has tablet table. Query describe_ring with the table parameter!", keyspace));
+    }
+    co_return co_await locator::describe_ring(_db.local(), _gossiper, keyspace, include_only_local_dc);
+}
+
+future<std::vector<dht::token_range_endpoints>>
+storage_service::describe_ring_for_table(const sstring& keyspace_name, const sstring& table_name) const {
+    slogger.debug("describe_ring for table {}.{}", keyspace_name, table_name);
+    auto& t = _db.local().find_column_family(keyspace_name, table_name);
+    if (!t.uses_tablets()) {
+        auto ranges = co_await describe_ring(keyspace_name);
+        co_return ranges;
+    }
+    table_id tid = t.schema()->id();
+    auto erm = t.get_effective_replication_map();
+    auto& tmap = erm->get_token_metadata_ptr()->tablets().get_tablet_map(tid);
+    const auto& topology = erm->get_topology();
+    std::vector<dht::token_range_endpoints> ranges;
+    co_await tmap.for_each_tablet_gently([&] (locator::tablet_id id, const locator::tablet_info& info) -> future<> {
+        co_await coroutine::maybe_yield();
```

Comment:
If that's the only yield point, you could use the existing for_each_tablet() which yields in the same place.

Reply:
I think the version that the callback returns a future is better.  In moved the yield to the for loop two.

Reply:
Agree

Reply:
The new version does not have explicit yield in each loop in for_each_tablet anymore. The call to the future<> func() is already a yield point. Why do we need `co_await coroutine::maybe_yield()' in each loop?

```
future<> tablet_map::for_each_tablet(seastar::noncopyable_function<future<>(tablet_id, const tablet_info&)> func) const {
    std::optional<tablet_id> tid = first_tablet();
    for (const tablet_info& ti : tablets()) {
        co_await func(*tid, ti);
        tid = next_tablet(*tid);
    }   
} 
```

Reply:
IIUC, if the said `func()` returns an available future,  in non-debug build, seastar does not schedule it, in other words, it does not yield in that case.

Reply:
co_await already checks for preemption, I think:
```
    bool await_ready() const noexcept {
        return _future.available() && (!CheckPreempt || !need_preempt());
    }
```
So we don't need an extra yield check in for_each_tablet().

> Why do we need `co_await coroutine::maybe_yield()' in each loop?

We don't.

Reply:
ahh, right! sorry, i missed the `co_await` part, indeed.